src/SimpleSAML/Metadata/MetaDataStorageHandler.php

```diff
@@ -264,32 +264,46 @@ public function getPreferredEntityIdFromCIDRhint(string $set, string $ip): ?stri
      */
     public function getMetaDataForEntities(array $entityIds, string $set): array
     {
+        $startTime = microtime(true);
         $result = [];
+        // We are flipping the entityIds array in order to avoid constant iteration over it.
+        // Even if it becomes smaller over time.
+        // Still, after flipping all actions will be O(1)
+        $entityIdsFlipped = array_flip($entityIds);
         $timeUtils = new Utils\Time();
+
         foreach ($this->sources as $source) {
-            $srcList = $source->getMetaDataForEntities($entityIds, $set);
+            // Break iteration if the entityID List is empty
+            if (empty($entityIdsFlipped)) {
+                break;
+            }
+
+            $srcList = $source->getMetaDataForEntities($entityIdsFlipped, $set);
             foreach ($srcList as $key => $le) {
-                if (array_key_exists('expire', $le)) {
-                    if ($le['expire'] < time()) {
-                        unset($srcList[$key]);
-                        Logger::warning(
-                            "Dropping metadata entity " . var_export($key, true) . ", expired " .
-                            $timeUtils->generateTimestamp($le['expire']) . ".",
-                        );
-                        continue;
-                    }
+                if (!empty($le['expire']) && $le['expire'] < time()) {
+                    unset($srcList[$key]);
+                    Logger::warning(
+                        'Dropping metadata entity ' . var_export($key, true) . ', expired ' .
+                        $timeUtils->generateTimestamp($le['expire']) . '.',
+                    );
+                    continue;
                 }
                 // We found the entity id so remove it from the list that needs resolving
                 /** @psalm-suppress PossiblyInvalidArrayOffset */
-                unset($entityIds[array_search($key, $entityIds)]);
+                unset($entityIdsFlipped[$key]);
+                // Add the key to the result set
+                $result[$key] = $le;
             }
-            $result = array_merge($srcList, $result);
         }
 
+        $endTime = microtime(true);
+        $executionTime = $endTime - $startTime;
+        $formattedTime = number_format($executionTime, 6, '.', '');
+        Logger::info(__METHOD__ . "Execution time: $formattedTime seconds");
```

Comment:
I'd prefer not seeing the timing included and logged as part of the PR.     Overall I like the idea of being able to get timing information out of SSP on how long different things takes `authproc`, metadata retrieval,  saving/reading from the store, etc  but I think we should do that in a different conversation and get some more input from community. For example we have thresholds where that log level for timings adjusts based on how long it takes: e.g. fast is debug, slower is info, etc.      So my vote would be for removing timing from this PR and continuing the idea of how to do timing and threshold configuration in slack or github.

Reply:
@pradtke i added the time tracking for testing purposes while validating the fix. I will remove it as soon as we have a consensus on the implementation.